src/poetry_plugin_freeze/app.py

```diff
@@ -121,15 +124,82 @@ def get_dep_packages(self):
                 project_requires=root_package.all_requires,
                 root_package_name=root_package.name,
                 project_python_marker=root_package.python_marker,
+                extras=root_package.extras,
             )
         )
         return {p.package.name: p for p in dep_packages}
 
+    @lru_cache(maxsize=None)
+    def get_dependency_sources(self):
+        """Determine the root source of each locked dependency
+
+        For each locked dependency, determine whether it came
+        as a base requirement or part of one or more extras.
+        """
+
+        def _with_python_marker(requirements, root_package):
+            """Augment requirements with the root package's python marker"""
+            marked_requirements = []
+            for require in requirements:
+                require = require.clone()
+                require.marker = require.marker.intersect(root_package.python_marker)
+                marked_requirements.append(require)
+            return marked_requirements
+
+        repository = self.poetry.locker.locked_repository()
+        root_package = self.poetry.package
+        locked_packages_by_name = {p.name: [p] for p in repository.packages}
+        dependency_sources = {}
+
+        # Identify nested dependencies that don't require extra selections.
+        base_nested_dependencies = walk_dependencies(
+            _with_python_marker(root_package.all_requires, root_package),
+            packages_by_name=locked_packages_by_name,
+            root_package_name=root_package.name,
+        )
+        for d in base_nested_dependencies:
+            dependency_sources.setdefault(d.name, set()).add("base")
```

Comment:
why the add base if its just going to be removed? just leave it an empty set.

Reply:
Thanks for pointing this out, I wanted to make sure that we don't freeze an `extra` marker for dependencies if they come in through both the base dependency set and extras.

...but I didn't actually have the code or tests in place to handle that case! Just added now.